### PR TITLE
charts/gateway OTK Release 4.6.2

### PR DESCRIPTION
**Pull request by APIM-KiranVaddadi**
**Description of the change**
 
  - OTK DB install/upgrade using Liquibase scripts for MySql and Oracle.
  - OTK DB install/upgrade on the gateways MySQL container (MySQL subchart)
  - Install/upgrade OTK of type SINGLE on Ephemeral gateways using initContainer is now supported.
  - Added OTK Connection properties to support c3p0 settings.
  - Added support OTK read-only connections for MySQL and Oracle.
  - Added support for OTK policies customization through config maps and secrets.
  - OTK DMZ/Internal gateway certs can now be configured using values file.

**Benefits**

**Drawbacks**

**Applicable issues**

  - fixes #

**Additional information**

**Checklist** 
- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[charts/gateway]`)
- [x] If the chart contains a `values-production.yaml` apart from `values.yaml`, ensure that you implement the changes in both files

### CODE REVIEW

File: charts/gateway/production-values.yaml [hunk 1]
```diff
@@ -485,36 +485,110 @@ service:
...
+    # Specify Internal gateway certificates that needs to be imported for DMZ GW (for ephemeral gateway)
+    # internalGatewayCert:
+    # internalGatewayIssuer:
+    # internalGatewaySerial:
+    # internalGatewaySubject:
+
+  # List of comma separated sub solution Kits to install or upgrade.
   # subSolutionKitNames:
 
+  customizations:
```
Comment (Gazza7205) [inline charts/gateway/production-values.yaml:539]: Customizations shouldn't use the file path method - (540-542)

Using a path is generally not a good idea for a few reasons (this was the original way we loaded bundles and is only retained for legacy)
- Requires you clone the repo to set that path and bundles up
- Becomes part of the Helm Release manifest which is bound to the 1mb size limit for the entire release so could be a breaking change
- Makes upgrading difficult

The documentation should focus more on how to create the customizations bundle vs. reference it (existingBundle or initContainer)

File: charts/gateway/templates/otk-scheduled-task-jobs.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            {{- toYaml $.Values.otk.job.podLabels | nindent 12}}
+        {{- end }}
+        {{- if  $.Values.otk.job.podAnnotations }}
+          annotations: {{- toYaml $.Values.otk.job.podAnnotations | nindent 12 }}
+        {{- end }}
+        spec:
+          {{- if $.Values.podSecurityContext }}
+          securityContext: {{- toYaml $.Values.podSecurityContext | nindent 12 }}
+          {{- end }}
+          containers:
```
Comment (Gazza7205) [inline charts/gateway/templates/otk-scheduled-task-jobs.yaml:46]: Image and imagePullSecret(s) should be configurable in values.yaml

File: charts/gateway/values.yaml [hunk 3]
```diff
@@ -485,36 +485,111 @@ service:
...
+    # Specify Internal gateway certificates that needs to be imported for DMZ GW (for ephemeral gateway)
+    # internalGatewayCert:
+    # internalGatewayIssuer:
+    # internalGatewaySerial:
+    # internalGatewaySubject:
+
+  # List of comma separated sub solution Kits to install or upgrade.
   # subSolutionKitNames:
 
+  customizations:
```
Comment (Gazza7205) [inline charts/gateway/values.yaml:539]: Same comment as production-values.yaml

File: charts/gateway/values.yaml [hunk 4]
```diff
@@ -537,38 +612,76 @@ otk:
...
-    password: 7layer
-    # properties: {"maximumPoolSize":15, "minimumPoolSize":3}
+    connectionName: OAuth
+    dbUpgrade: true
+    useDemoDb: true
+
+    username: otk_user
+    password: mypassword
+
+    # ddlUsername:
```
Comment (Gazza7205) [inline charts/gateway/values.yaml]: ddlUsername and ddlPassword need more context - this is a database user that has the ability to create databases (i.e. admin?).

values that are not relevant to both sql and cassandra should be split into the database.sql and database.cassandra sections

File: examples/otk/README.md [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    properties: {"localDataCenterName" : "dc1"}
+    cassandra:
+      connectionPoints: cassandra-c1,cassandra-c2
+      port: 9042
+      keyspace: otk_db
+....
+.....
+```
+
+### Customizations
```
Comment (Gazza7205) [inline examples/otk/README.md:166]: Same comment as values-production.yaml

This should give more context to what OTK Customizations are, how to extract them, how to customize them, etc.

This should also extend to graphman and the graphman-client which makes the resulting bundle much easier to read, configure and add to VCS

File: charts/gateway/templates/otk-install-configmap.yaml [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+{{- end }}
+data:
+  OTK_TYPE: {{ template "otk-install-type" .}}
+  {{- if .Values.database.enabled }}
+  OTK_INSTALL_MODE: "job"
+  {{- else }}
+  OTK_INSTALL_MODE: "initContainer"
+  {{- end }}
+  BOOTSTRAP_DIR: "/opt/SecureSpan/Gateway/node/default/etc/bootstrap/bundle/000OTK"
+  OTK_SK_UPGRADE: "true"
```
Comment (Gazza7205) [inline charts/gateway/templates/otk-install-configmap.yaml:28]: Several hardcoded values - should be configurable

File: charts/gateway/values.yaml [hunk 7]
```diff
@@ -485,36 +485,111 @@ service:
...
 
-  # Force install or upgrade by uninstalling existing otk soluction kit and install.
-  # forceInstallOrUpgrade: false
+  # Force install or upgrade by uninstalling existing otk solution kit and install.
+  forceInstallOrUpgrade: false
 
   # Not applicable for DMZ and INTERNAL OTK types
-  # enablePortalIngeration: false
-  # skipPostInstallationTasks: false
+  enablePortalIntegration: false
```
Comment (Gazza7205) [inline charts/gateway/values.yaml:499]: Assuming this is the Portal Integration Components and not actually enrolling with Portal?

This should be clarified

File: charts/gateway/templates/otk-scheduled-task-jobs.yaml [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+          - name: schedule-task
+            image: curlimages/curl:7.72.0
+            {{- if $.Values.containerSecurityContext }}
+            securityContext: {{- toYaml $.Values.containerSecurityContext | nindent 14 }}
+            {{- end }}
+            imagePullPolicy: IfNotPresent
+            command:
+            - /bin/sh
+            - -c
+            - "curl -v --request POST 'https://{{ template "gateway.fullname" $ }}:8443/oauth/db-maintenance?task={{ .name }}' --insecure"
```
Comment (Gazza7205) [inline charts/gateway/templates/otk-scheduled-task-jobs.yaml]: is there any auth on this endpoint? what's the risk of a bad actor running these tasks?

This should be accessible on management ports only

Update: Restricts based on Address and is not configurable/documented - this needs to be addressed

**Comment by Gazza7205** [comment]
In DMZ and Internal mode changes are made to the customizations folder - these changes would be overritten by a customization bundle applied at bootstrap or via Restman/Graphman.

We should document which policies are changed per mode and advise excluding them from custom customization bundles.

This applies to the following (varying across otk types)
- `#OTK OVP Configuration`
- `#OTK Storage Configuration`
- `#OTK  Client Context Variables`
- OTK FIP Client Authentication Extension

File: charts/gateway/production-values.yaml [hunk 9]
```diff
@@ -538,37 +612,75 @@ otk:
...
+    useDemoDb: true
+
+    username: otk_user
+    password: mypassword
+
+    # ddlUsername:
+    # ddlPassword:
+    # existingSecretName: otkDbSecret
+
     properties: {"maximumPoolSize":15, "minimumPoolSize":3}
```
Comment (Gazza7205) [inline charts/gateway/production-values.yaml:622]: otk.database.properties, otk.database.sql.connectionProperties and otk.database.cassandra.driverConfig appear to be ambiguous. Could you add some more clarification on which one to use?

File: charts/gateway/templates/otk-install-configmap.yaml [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+  {{- end }}
+  {{- if eq .Values.otk.database.type "mysql" }}
+  {{- if (.Values.otk.database.useDemoDb)}}
+  OTK_JDBC_URL: jdbc:mysql://{{ .Release.Name }}-mysql:3306/{{ default "otk_db" .Values.otk.database.sql.databaseName }}
+  {{- else }}
+  OTK_JDBC_URL: {{ required "Please fill in otk.database.sql.jdbcURL in values.yaml" .Values.otk.database.sql.jdbcURL | quote }}
+  {{- end }}
+  {{- end }}
+  {{- end }}
+  OTK_UPDATE_DATABASE_CONNECTION: {{default false .Values.otk.database.update | quote }}
```
Comment (Gazza7205) [inline charts/gateway/templates/otk-install-configmap.yaml]: This is otk.database.updateConnection in values.yaml and the readme

File: charts/gateway/values.yaml (hunk 4 above)
Comment (APIM-KiranVaddadi) [reply charts/gateway/values.yaml]: Added in line comments. moved to database.sql. Not applicable for cassandra.

File: charts/gateway/production-values.yaml (hunk 1 above)
Comment (APIM-KiranVaddadi) [reply charts/gateway/production-values.yaml:539]: Removed the configuration from values and values production. Templates unchanged as this option is used by System team.

File: charts/gateway/values.yaml (hunk 3 above)
Comment (APIM-KiranVaddadi) [reply charts/gateway/values.yaml:539]: Removed the configuration from values and values production. Templates unchanged as this option is used by System team.

File: examples/otk/README.md (hunk 5 above)
Comment (APIM-KiranVaddadi) [reply examples/otk/README.md:166]: Detailed explanation is given at /examples/otk and will expanded as needed.

File: charts/gateway/templates/otk-install-configmap.yaml (hunk 6 above)
Comment (APIM-KiranVaddadi) [reply charts/gateway/templates/otk-install-configmap.yaml:28]: Hard coded to have a predictable loading of the policies.

File: charts/gateway/values.yaml (hunk 7 above)
Comment (APIM-KiranVaddadi) [reply charts/gateway/values.yaml:499]: Yes. Updated the docs.

File: charts/gateway/production-values.yaml (hunk 9 above)
Comment (APIM-KiranVaddadi) [reply charts/gateway/production-values.yaml:622]: database.properties are generic connection properties
connection properties are c3p0 properties
driverConfig properties are Cassandra driver properties 
Updated docs

File: charts/gateway/templates/otk-scheduled-task-jobs.yaml (hunk 8 above)
Comment (APIM-KiranVaddadi) [reply charts/gateway/templates/otk-scheduled-task-jobs.yaml]: Its now configurable

File: charts/gateway/templates/otk-install-configmap.yaml (hunk 10 above)
Comment (APIM-KiranVaddadi) [reply charts/gateway/templates/otk-install-configmap.yaml]: Updated.

File: charts/gateway/templates/otk-scheduled-task-jobs.yaml (hunk 2 above)
Comment (APIM-KiranVaddadi) [reply charts/gateway/templates/otk-scheduled-task-jobs.yaml:46]: Added configurable image pull secret.

**Review by Gazza7205 — APPROVED** [review]

**Pull request merged**